java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightServerTestRule.java

```diff
@@ -170,18 +134,88 @@ private FlightServer getStartServer(Function<Location, FlightServer> newServerFr
     throw new IOException(exceptions.pop().getCause());
   }
 
-  private CallHeaderAuthenticator.AuthResult validate(final String username,
-                                                      final String password) {
-    if (validateUser(username, password)) {
-      return () -> username;
-    }
-
-    throw CallStatus.UNAUTHENTICATED.withDescription("Invalid credentials.").toRuntimeException();
-  }
-
   @Override
   public void close() throws Exception {
     allocator.getChildAllocators().forEach(BufferAllocator::close);
     AutoCloseables.close(allocator);
   }
+
+  /**
+   * Builder for {@link FlightServerTestRule}.
+   */
+  public static final class Builder {
+    private final Properties properties = new Properties();
+    private FlightSqlProducer producer;
+    private Authentication authentication;
+
+    /**
+     * Sets the host for the server rule.
+     * @param host the host value.
+     * @return     the Builder.
+     */
+    public Builder host(final String host) {
+      properties.put(ArrowFlightConnectionConfigImpl.ArrowFlightConnectionProperty.HOST.camelName(), host);
+      return this;
+    }
+
+    /**
+     * Sets a random port to be used by the server rule.
+     * @return the Builder.
+     */
+    public Builder randomPort() {
+      properties.put(ArrowFlightConnectionConfigImpl.ArrowFlightConnectionProperty.PORT.camelName(),
+          FreePortFinder.findFreeLocalPort());
+      return this;
+    }
+
+    /**
+     * Sets a specific port to be used by the server rule.
+     * @param port  the port value.
+     * @return      the Builder.
+     */
+    public Builder port(final int port) {
+      properties.put(ArrowFlightConnectionConfigImpl.ArrowFlightConnectionProperty.PORT.camelName(), port);
+      return this;
+    }
+
+    /**
+     * Sets the producer that will be used in the server rule.
+     * @param producer  the flight sql producer.
+     * @return          the Builder.
+     */
+    public Builder producer(final FlightSqlProducer producer) {
+      this.producer = producer;
+      return this;
+    }
+
+    /**
+     * Sets the type of the authentication that will be used in the server rules.
+     * There are two types of authentication: {@link UserPasswordAuthentication} and
+     * {@link TokenAuthentication}.
+     * @param authentication  the type of authentication.
+     * @return                the Builder.
+     */
+    public Builder authentication(final Authentication authentication) {
+      this.authentication = authentication;
+      return this;
+    }
+
+    /**
+     * Builds the {@link FlightServerTestRule} using the provided values.
+     * @return  a {@link FlightServerTestRule}.
+     */
+    public FlightServerTestRule build() {
+      if (authentication instanceof UserPasswordAuthentication) {
```

Comment:
Can we refactor it so we don't need to use `instanceof`? Maybe add a getProperties() to the Authentication interface

Reply:
Fixed